src/irmin/store_intf.ml

```diff
@@ -875,18 +908,46 @@ module type S = sig
     [> write ] Private.Contents.t ->
     [> read_write ] Private.Node.t ->
     tree ->
-    hash Lwt.t
+    kinded_key Lwt.t
   (** Save a tree into the database. Does not do any reads. If [clear] is set
       (it is by default), the tree cache will be cleared after the save. *)
 end
 
-module type Maker = sig
+module type S = sig
+  type hash
+
+  (** @inline *)
+  include
+    S_generic_key
+      with type Schema.Hash.t = hash
+       and type hash := hash
+       and type contents_key = hash
+       and type node_key = hash
+       and type commit_key = hash
+end
```

Comment:
I tried making `S` the more general form, and having e.g. `Hash_keyed` for the simple case, but it's a big backwards-compatibility liability since any occurrences of `Irmin.S` in the wild (e.g. as functor arguments) are likely to make the assumption that keys are hashes.

I settled on adding a new `Generic_key.S` as it introduces less churn.